syz-manager/repro_test.go

```diff
@@ -105,9 +108,55 @@ func TestReproOrder(t *testing.T) {
 	assert.Equal(t, crashes[2], obj.popCrash())
 }
 
+func TestReproRWRace(t *testing.T) {
```

Comment:
Now that the PR follows a different approach, I think the test should also change. What we essentially want to verify is that we don't reproduce bugs twice. So maybe something along these lines:

* We tell `needRepro` to return true.
* We `Enqueue` some `BugA` twice.
* As a part of the "reproduction process", we tell `needRepro` to start returning false.
* `Empty()` gets down to 0 (which means it didn't try to schedule another bug).

Reply:
Thanks for the comment. I've changed the unit test. I wanted to cover also the case when `reproMgr.onlyOnce = false` but AFAIU, in this case the code allows to `Enqueue` crash with the same title twice. Hence, 2 reproducers are scheduled which is not the expected case per description. Currently, it's set only here via `mgr.cfg.DashboardOnlyRepro` flag.

https://github.com/google/syzkaller/blob/9f0ab3fb8930510928e013275e4849b621d82efb/syz-manager/manager.go#L316

Please let me know comments about the current patch. I've desperately would like to cover with interfaces and unit tests syz-manager in next PRs with dependency injection mechanism.

Reply:
Thanks for addressing the comments!

Syzkaller should never reproduce two crashes with the same title simultaneously - it should always do it sequentially. If `onlyOnce` is true, it will never schedule the second reproduction, regardless of the previous reproduction result.

I think that we'd rather want to verify the case when it does the second reproduction attempt (at least by calling `needRepro` and seeing it return false), so `onlyOnce`'d better be `false`.